helix-dap/src/client.rs

```diff
@@ -0,0 +1,647 @@
+#[serde(rename_all = "camelCase")]
+pub struct OutputEventBody {
+    pub output: String,
+    pub category: Option<String>,
+    pub group: Option<String>,
+    pub line: Option<usize>,
+    pub column: Option<usize>,
+    pub variables_reference: Option<usize>,
+    pub source: Option<Source>,
+    pub data: Option<Value>,
+}
+
+#[derive(Debug, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct StoppedEventBody {
+    pub reason: String,
+    pub description: Option<String>,
+    pub thread_id: Option<usize>,
+    pub preserve_focus_hint: Option<bool>,
+    pub text: Option<String>,
+    pub all_threads_stopped: Option<bool>,
+    pub hit_breakpoint_ids: Option<Vec<usize>>,
+}
+
+#[derive(Debug)]
+pub struct Client {
+    id: usize,
+    _process: Option<Child>,
+    server_tx: UnboundedSender<Request>,
+    request_counter: AtomicU64,
+    capabilities: Option<DebuggerCapabilities>,
+    awaited_events: Arc<Mutex<HashMap<String, Sender<Event>>>>,
+}
+
+impl Client {
+    pub fn streams(
+        rx: Box<dyn AsyncBufRead + Unpin + Send>,
+        tx: Box<dyn AsyncWrite + Unpin + Send>,
+        id: usize,
+        process: Option<Child>,
+    ) -> Result<Self> {
+        let (server_rx, server_tx) = Transport::start(rx, tx, id);
+
+        let client = Self {
+            id,
+            _process: process,
+            server_tx,
+            request_counter: AtomicU64::new(0),
+            capabilities: None,
+            awaited_events: Arc::new(Mutex::new(HashMap::default())),
+        };
+
+        tokio::spawn(Self::recv(Arc::clone(&client.awaited_events), server_rx));
+
+        Ok(client)
+    }
+
+    async fn recv(
+        awaited_events: Arc<Mutex<HashMap<String, Sender<Event>>>>,
+        mut server_rx: UnboundedReceiver<Payload>,
+    ) {
+        while let Some(msg) = server_rx.recv().await {
+            match msg {
+                Payload::Event(ev) => {
+                    let name = ev.event.clone();
+                    let hashmap = awaited_events.lock().await;
+                    let tx = hashmap.get(&name);
+
+                    match tx {
+                        Some(tx) => match tx.send(ev).await {
+                            Ok(_) => (),
+                            Err(_) => error!(
+                                "Tried sending event into a closed channel (name={:?})",
+                                name
+                            ),
+                        },
+                        None => {
+                            info!("unhandled event");
+                            // client_tx.send(Payload::Event(ev)).expect("Failed to send");
+                        }
+                    }
+                }
+                Payload::Response(_) => unreachable!(),
+                Payload::Request(_) => todo!(),
+            }
+        }
+    }
+
+    pub async fn listen_for_event(&self, name: String) -> Receiver<Event> {
```

Comment:
This needs either an extra hashmap for one-time event handlers or a method to remove old event receivers, because in some cases DAP sessions might live long enough to care about memory leaks in them.

Reply:
Outdated, now handling those in event loop.